lib/spack/spack/stage.py

```diff
@@ -415,10 +418,25 @@ def expanded(self):
         """Returns True if source path expanded; else False."""
         return os.path.exists(self.source_path)
 
+    @property
+    def path(self):
+        return str(self._path)
+
     @property
     def source_path(self):
         """Returns the well-known source directory path."""
-        return os.path.join(self.path, _source_path_subdir)
+        return str(self._path / _source_path_subdir)
+
+    @property
+    def build_directory(self):
+        """Returns a potential build directory, defaulting to in source build"""
+        return self.source_path
```

Comment:
I'd prefer if the base `Stage` class did not store a notion of `build_directory`, but I could be overlooking some complexities of handling this specifically in `CMakeBuildStage`